docs/intro/intro_symbolic_math.rst

```diff
@@ -0,0 +1,85 @@
+.. _intro_symbolic_math:
+
+=============
+Symbolic Math
+=============
+
+dimod enables easy incorporation of binary and integer variables as
+:ref:`single-variable models <symbolic_math_generators>`. For example, you can
+represent such binary variables as follows:
+
+>>> from dimod import Binary, Spin
+>>> x = Binary('x')
+>>> s = Spin('s')
+>>> x
+BinaryQuadraticModel({'x': 1.0}, {}, 0.0, 'BINARY')
+
+Similarly for integers:
+
+>>> from dimod import Integer
+>>> i = Integer('i')
+>>> i
+QuadraticModel({'i': 1.0}, {}, 0.0, {'i': 'INTEGER'}, dtype='float64')
+
+The construction of such variables as either BQMs or QMS depends on the type of
+variable:
+
+>>> x + s
+QuadraticModel({'x': 1.0, 's': 1.0}, {}, 0.0, {'x': 'BINARY', 's': 'SPIN'}, dtype='float64')
+>>> 3*i - x
+QuadraticModel({'i': 3.0, 'x': -1.0}, {}, -0.0, {'i': 'INTEGER', 'x': 'BINARY'}, dtype='float64')
+
+You can express mathematical functions on these variables using Python functions such
+as :func:`sum`\ [#]_\ :
+
+.. [#]
+  See the `Example: Adding Models`_ example for a performant summing function.
+
+>>> sum([3*i, 2*i])
+QuadraticModel({'i': 5.0}, {}, 0.0, {'i': 'INTEGER'}, dtype='float64')
+
+.. note::
+  It's important to remember that, for example, :code:`x = dimod.Binary('x')`
+  instantiates a single-variable model, in this case a
+  :class:`~dimod.binary.binary_quadratic_model.BinaryQuadraticModel` with
+  variable label ``'x'``, not a free-floating variable labeled ``x``. Consequently,
+  you can add ``x`` to another model, say :code:`bqm = dimod.BinaryQuadraticModel('BINARY')`,
+  by adding the two models, :code:`x + bqm`. This adds the variable labeled ``'x'``
+  in the single-variable BQM, ``x`` to model ``bqm``. You cannot add ``x`` to a
+  model---as though it were variable ``'x'``---by doing :code:`bqm.add_variable(x)`.
+
+Example: BQM
+============
+
+This example creates the BQM :math:`x + 2y -xy`:
+
+>>> from dimod import Binary
+>>> x = Binary('x')
+>>> y = Binary('y')
+>>> bqm = x + 2*y - x*y
+
+Example: CQM
+============
+
+This example uses symbolic math to set an objective (:math:`2i - 0.5ij + 10`)
+and constraints (:math:`xj <= 3` and :math:`i + j >= 1`) in a simple CQM.
+
+>>> from dimod import Binary, Integer, ConstrainedQuadraticModel
+>>> x = Binary('x')
+>>> i = Integer('i')
+>>> j = Integer('j')
+>>> cqm = ConstrainedQuadraticModel()
+>>> cqm.set_objective(2*i - 0.5*i*j + 10)
+>>> cqm.add_constraint(x*j <= 3)                   # doctest: +IGNORE_RESULT
+>>> cqm.add_constraint(i + j >= 1)                 # doctest: +IGNORE_RESULT
+
+Example: Adding Models
+======================
+
+This example uses the performant :func:`~dimod.quicksum` to add multiple models.
+
+>>> import random
+>>> from dimod import Binary, quicksum
+>>> vars = ['x'+str(i) for i in range(5)]
+>>> models = [Binary(var, bias=random.randint(0,5)) for var in vars]
```

Comment:
```python
>>> models = [v*random.randint(0, 5) for v in dimod.Binaries(variables)]
```

Reply:
Can also just do `dimod.BinaryArray(range(5))*np.random.randint(0, 5, size=5)` which IMO is the flow we actually want to be encouraging.

Reply:
I think showing both preferred patterns:
```python
In [14]: num_variables = 100

In [15]: %timeit (dimod.BinaryArray(range(num_variables))*np.random.randint(0, 5, size=num_variables)).sum()
2.11 ms ± 11.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [16]: %timeit dimod.quicksum(v*np.random.randint(0, 5) for v in dimod.Binaries(range(num_variables)))
2.5 ms ± 33.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
would be good (don't need the timing info, but it's interesting)

Reply:
For posterity, the latter ends up being faster
```
In [17]: num_variables = 5000

In [18]: %timeit (dimod.BinaryArray(range(num_variables))*np.random.randint(0, 5, size=num_variables)).sum()
287 ms ± 1.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [19]: %timeit dimod.quicksum(v*np.random.randint(0, 5) for v in dimod.Binaries(range(num_variables)))
128 ms ± 1.43 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
Do we still consider these as useful primarily for testing and small problems or have they begun moving into (high) performance now that we have also `BinaryArray`, `quicksum`, maybe the older single-var BQMs are more efficient too?

Reply:
We still don't encourage any symbolic for true performance, but with quicksum etc the performance can be OK.